### PR TITLE
Cleaning up makefiles. Fixes the makefile issues.

### PR DESCRIPTION
**Pull request by syclik**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary

This pull request simplifies the build and gets the makefiles in order. The makefiles got tangled up through a few recent changes. This uses changes from the PR in Math: stan-dev/math#956.

Here are some of the highlights:
- `CXX` is the only compiler setting used. In develop, both `CXX` and `CC` have to be set
- Windows: removed the `uname: not found` message
- `CXXFLAGS` are now broken into smaller components that can be set by the user:
    - `CXXFLAGS_LANG` for language-specific flags
    - `CXXFLAGS_WARNINGS` for ignoring warnings
    - `CXXFLAGS_BOOST` for boost-specific flags
    - `CXXFLAGS_SUNDIALS` for sundials-specific flags
    - `CXXFLAGS_EIGEN` for eigen-specific flags
    - `CXXFLAGS_GTEST` for google-test-specific flags
    - `CXXFLAGS_OS` for a catch-all and any os-specific flags.
- `CPPFLAGS` are broken into components (`CPPFLAGS` are the C preprocessor flags):
    - `CPPFLAGS_BOOST` for boost-specific flags
    - `CPPFLAGS_SUNDIALS` for sundials-specific flags
    - `CPPFLAGS_EIGEN` for eigen-specific flags
    - `CPPFLAGS_GTEST` for google-test-specific flags
    - `CPPFLAGS_OPENCL` for opencl

  breaking apart the flags this way makes it clearer how to extend the flags for MPI and GPU.
- each of the flags can be set within `make/local` and it will take precedence over the default values
- the dependency generation now generates with dummy targets for each file. This means more robustness in that -- this fixes problems where a header file is removed from the source tree. This caused a lot of trouble with stability of make.
- this should fix the dependencies with updating `function_signatures.h`
- we no longer have to update the `make/tests` folder with hand-coded dependencies to the generated `.hpp` file from `.stan` files! This will cut down error. Just need to `#include <test/test-models/good/*.hpp>` in the test file and it'll auto-generate!

#### Intended Effect

Cleans up a lot of the build. Gets us up and building for Windows too.

#### How to Verify

Run all the tests and other things we've done before:
```
./runTests.py src/test/unit
./runTests.py src/test/integration
make doc
make
```

#### Side Effects

These changes will affect CmdStan. We'll need to update that repo too.

#### Documentation

- [x] This needs to be documented on the wiki and in the installation instructions. We'll see if we can get the build to pass first.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

Generable

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bbbales2** [comment]
Sick hash @ fff1fff

**Comment by syclik** [comment]
Woohoo! I've got Stan working now.

**Comment by syclik** [comment]
It's me again. I have all the makefiles working for Math, Stan, and CmdStan! Mind checking this out and seeing if you spot any problems?

@bbbales2, can you specifically check to see that if you run a test, update the parser, then run the test again, it'll work appropriately?

@mitzimorris, I was able to kill having to write out dependencies in the makefile for tests if it's in the `#include` of the test itself! Can you make sure that's working?

cc: @seantalts @bbbales2

**Comment by mitzimorris** [comment]
thanks - that worked.  will continue code review

**Comment by seantalts** [comment]
This PR was caught in a seemingly infinite loop on Jenkins. I killed it after 16 hours: http://d1m1s1b1.stat.columbia.edu:8080/blue/organizations/jenkins/Stan/detail/PR-2601/43/pipeline/47

**Comment by syclik** [comment]
Thanks! I'll dig into why it got stuck.

**Comment by mitzimorris** [comment]
I got the same thing running locally:

```
chmod 555 src/test/test-models/bad/read_only
clang++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0     -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE  -c -MT test/performance/logistic_test.o -MT src/test/performance/logistic_test.d -MM -MG -MP -MF src/test/performance/logistic_test.d src/test/performance/logistic_test.cpp
chmod 444 src/test/test-models/*/stanc_helper.stan
chmod 555 src/test/test-models/bad/read_only
clang++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0     -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE  -c -MT test/performance/logistic_test.o -MT src/test/performance/logistic_test.d -MM -MG -MP -MF src/test/performance/logistic_test.d src/test/performance/logistic_test.cpp
chmod 444 src/test/test-models/*/stanc_helper.stan
chmod 555 src/test/test-models/bad/read_only
clang++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0     -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE  -c -MT test/performance/logistic_test.o -MT src/test/performance/logistic_test.d -MM -MG -MP -MF src/test/performance/logistic_test.d src/test/performance/logistic_test.cpp
chmod 444 src/test/test-models/*/stanc_helper.stan
chmod 555 src/test/test-models/bad/read_only
clang++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0     -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE  -c -MT test/performance/logistic_test.o -MT src/test/performance/logistic_test.d -MM -MG -MP -MF src/test/performance/logistic_test.d src/test/performance/logistic_test.cpp
chmod 444 src/test/test-models/*/stanc_helper.stan
chmod 555 src/test/test-models/bad/read_only
clang++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0     -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE  -c -MT test/performance/logistic_test.o -MT src/test/performance/logistic_test.d -MM -MG -MP -MF src/test/performance/logistic_test.d src/test/performance/logistic_test.cpp
chmod 444 src/test/test-models/*/stanc_helper.stan
chmod 555 src/test/test-models/bad/read_only
```

**Comment by mitzimorris** [comment]
this was running: 
```
make test/integration/compile_models_test
```

why rename `make test/integration/compile_models` to `make test/integration/compile_models_test` ?

**Comment by mitzimorris** [comment]
`./runTests.py src/test/unit`
is also getting into some kind of infinite loop.

**Comment by syclik** [comment]
@mitzimorris, the infinite loop should be gone now. Sorry about that.

**Comment by syclik** [comment]
@mitzimorris, want to check it again? I think it's ready to go.

**Comment by syclik** [comment]
@mitzimorris, to answer your question:
> why rename `make test/integration/compile_models` to `make test/integration/compile_models_test`?

Ah... that one. We somehow were generating executables that dropped the `_test`. That's really non-standard. It's sort of expected when you have `foo.cpp` to generate a `foo` executable. The same should hold for `compile_models_test.cpp`; that should generate `compile_models_test`, not `compile_models` as the executable.

So, to be consistent with everything else, it's been changed.

From the user's perspective, there's no difference if using `runTests.py`. Hope that makes sense.

**Comment by mitzimorris** [comment]
trying again, but this is failing:
```
./runTests.py src/test/integration

...
clang++    -c -x c -O0 -include lib/stan_math/lib/sundials_3.1.0/include/stan_sundials_printf_override.hpp -isystem lib/stan_math/lib/sundials_3.1.0/include lib/stan_math/lib/sundials_3.1.0/src/idas/idas_spils.c -o lib/stan_math/lib/sundials_3.1.0/src/idas/idas_spils.o
c++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION        -c src/test/test-models/stanc.cpp -o test/test-models/stanc.o
src/test/test-models/stanc.cpp:1:10: fatal error: 'stan/command/stanc_helper.hpp' file not found
#include <stan/command/stanc_helper.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
c++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION         -c -o test/stan/lang/grammars/expression_grammar_inst.o src/stan/lang/grammars/expression_grammar_inst.cpp
1 error generated.
make: *** [test/test-models/stanc.o] Error 1
make: *** Waiting for unfinished jobs....
src/stan/lang/grammars/expression_grammar_inst.cpp:1:10: fatal error: 'stan/lang/grammars/expression_grammar_def.hpp' file not found
#include <stan/lang/grammars/expression_grammar_def.hpp>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make: *** [test/stan/lang/grammars/expression_grammar_inst.o] Error 1
make -j1 test/integration/compile_models_test test/integration/compile_standalone_functions_test failed
exit now (09/12/18 14:53:54 CEST)
```

**Comment by syclik** [comment]
Blah. Ok. I'm on it.... just to confirm, this is on Mac? (I found
differences in behavior last night between Mac and Windows, so I had to
change some stuff around.)

**Comment by syclik** [comment]
@mitzimorris, I just ran this locally and didn't have any problems. Mind
checking that both Stan and Math are up to date? And... that both are clean
before starting?

**Comment by mitzimorris** [comment]
sorry, I keep forgetting to do the correct math-update call.
command `./runTests.py src/test/integration` works!

we need documentation on specific testing tasks for developers working on the Stan compiler,
specifically for:

- compiling a single model using the target `.hpp-test`
- testing the compiler/generator by compiling all targets via make task `test/integration/compile_models`

previously these worked directly via `make`.  now you need to use script `runTests.py`, correct?

**Comment by syclik** [comment]
> sorry, I keep forgetting to do the correct math-update call. command `./runTests.py src/test/integration works!`

No worries! This is one of those instances where a mono-repo would have been really, really helpful! I didn't want to move the submodules through these PRs, but that would have also made it easier.

> we need documentation on specific testing tasks for developers working on the Stan compiler

Yup. There's a thread on discourse that talks about it: https://discourse.mc-stan.org/t/running-single-model-integration-test/5574

I'm updating the Stan wiki page now: https://github.com/stan-dev/stan/wiki/Testing-Stan-using-Gnu-Make-and-Python

> previously these worked directly via `make`. now you need to use script `runTests.py`, correct?

Nope! They still work! There is a slight difference: if you want to test `src/test/test-models/good/funs1.stan`, the new make call is: `make test/test-models/good/funs1.hpp-test`

The difference: removing `src/` in front. There's a good reason for this change. Moving the generated hpp file to `test/test-models/good` (out of the `src/` directory) allows us to not have to add the generated files by hand in `make/tests` to the tests that depend on generated hpp files. I think it's well worth it to not track it!

**Comment by mitzimorris** [comment]
heads up - refactor for tuples just got merged, and there's conflicts.  in particular, because travis kept timing out on the stan/lang unit tests, I broke it down by subdirectory.  also, there's now a whole bunch more generator tests - file src/test/unit/lang/generator_test.cpp is gone and there's a directory src/test/unit/lang/generator which has lots of tests.

**Comment by mitzimorris** [comment]
> Nope! They still work! There is a slight difference: if you want to test `src/test/test-models/good/funs1.stan`, the new make call is: `make test/test-models/good/funs1.hpp-test`

this works - but two things:

-  without initial `src/` you lose auto-completion for the path to the file name
-  there's a lot of repeated "-I' in the command, e.g.:

```
clang++ -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I . -I . -I . -I . -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0 -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0 -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0 -I lib/stan_math/lib/gtest_1.7.0/include -I lib/stan_math/lib/gtest_1.7.0        -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION    -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_USE_OWN_TR1_TUPLE -DGTEST_USE_OWN_TR1_TUPLE  -c src/test/test-models/stanc.cpp -o test/test-models/stanc.o
```

**Comment by syclik** [comment]
Ok. Mind helping me out with the Jenkinsfile conflict? I'll look at the
tests.

**Comment by mitzimorris** [comment]
absolutely - will take care of `.travis.yml`

**Comment by syclik** [comment]
@mitzimorris: cool! I'll try to remove the duplicate includes. It doesn't
hurt, but I think I know how to remove them. And yes, it sucks that we
can't use autocomplete. I think we live with it for now. This is in
exchange for not having to manually list test dependencies like we used to.
We can see if we can do something about it later; I just don't know how to
do it now and not have to list those files.

**Comment by syclik** [comment]
Everything is building! Math, Stan, and CmdStan.

**Comment by mitzimorris** [comment]
tried running `make test-headers` - crashes:

```
X_NO_VARIADIC_EXPRESSION     -c -O0 -include src/stan/model/gradient.hpp test/dummy.cpp -o /dev/null
clang++ -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -std=c++1y -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare      -O0 -I src -I lib/stan_math/ -I lib/stan_math/lib/eigen_3.3.3 -I lib/stan_math/lib/boost_1.66.0 -I lib/stan_math/lib/sundials_3.1.0/include    -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION     -c -O0 -include src/stan/model/model_header.hpp test/dummy.cpp -o /dev/null
fatal error: file '/Users/mitzi/github/stan-dev/cmdstan/stan/src/stan/model/model_header.hpp' has been modified since the precompiled header 'src/stan/model/model_header.hpp.gch'
      was built
note: please rebuild precompiled header 'src/stan/model/model_header.hpp.gch'
1 error generated.
make: *** [src/stan/model/model_header.hpp-test] Error 1
make: *** Waiting for unfinished jobs....
```

**Comment by syclik** [comment]
Can you clean everything first?

**Comment by syclik** [comment]
I'm not sure how the precomputed headers we're generated originally.

**Comment by syclik** [comment]
I think the minimal thing to do is to:
rm src/stan/model/model_header.hpp.gch

I think an older CmdStan build has left that behind and it's not consistent.

**Comment by syclik** [comment]
Also... I tried with the current, up-to-date versions. If I build from
CmdStan and leave behind stan/src/stan/model/model_header.hpp.gch and I
jump into Stan, I don't see any issues.

Our Stan makefiles don't take advantage of the precompiled headers, but
that's something for a future issue.


Daniel

**Comment by mitzimorris** [comment]
for the language unit tests, I'm looking at the generated `.d` file and it lists the set of ast node files but it doesn't list the associated `_def.hpp` files.  should it also list these files?  if I change the behavoir of an ast node but not the signature, changes to the `_def.hpp` file aren't picked up automatically.

**Comment by mitzimorris** [comment]
actually, not asking `should it`, requesting that it do this, if at all reasonable.

**Comment by syclik** [comment]
@mizimorris, the behavior you're looking for should already happen. The
request for putting that info into those particular files isn't right; I'll
try to explain how we do it. (I'm on my phone and working from memory. I
can send more detailed info later if needed.)

First: how are you testing? It should work. Meaning that if you touch a
file that's necessary to build one of the explicit template instantiation
.o files, that will trigger the .o file to be rebuilt, the libstanc.a to be
rebuilt, then finally the stanc to be rebuilt. If I can reproduce it
locally, that'll help.

How we're doing it. See make/tests, around line 35. One of the ifs should
include files that are the dependency files of the
TEMPLATE_INSTANTIATION_CPP files. These are the files that contain what
you're looking for. We can't arbitrarily put additional dependencies into
the .d files since they're autogenerated by the c++ compiler. But, we've
already put rules that state that:
- certain tests depend on stanc
- stanc depends on all those .o files
- the .o files depend on .cpp files
- the dependency files indicate which files the .cpp files depend on

Anyway, example that doesn't work as expected? I can produce one that does
work.

**Comment by mitzimorris** [comment]
here's what I did as a test:

- clean everything

- run unit test `src/test/unit/lang/parser/rng_location_test.cpp`
which checks that file `src/test/test-models/good/rng_loc.stan` can be compiled.

-  in grammar file `src/stan/lang/grammars/semantic_actions_def.hpp` changed one of the semantic actions called during the parser of `rng_loc.stan` so that it prints to std::out when called.

- reran test, confirmed that change to semantic_actions_def was picked up - unit test run contains print stmt.

- in file ast/node/fun_def.hpp made a trivial change to the constructor so that it prints to std::out.  changed program `rng_loc.stan` adding statements to the Stan program so that the parser will have to call the constructor.  (e.g.:  `vector[3] x; vector[3]y = x + 4*x;`).

- reran test, change to `fun_def.hpp` not picked up; unit test output remains the same.

- ran `make clean-all`,  then ran test, now test output has print from call to constructor in `ast/node/fun_def.hpp`.

**Comment by syclik** [comment]
Thank you. I'll try to see what's missing. I'll let you know what I do to
fix it.

**Comment by mitzimorris** [comment]
not fixing it for this PR is OK - filing an issue is good enough - I can provide more details to show how to test if you need them.

**Comment by syclik** [comment]
@mitzimorris, I tried your test, but it works on my machine. Would you mind jumping on a video call where you can screen share?

File: make/tests [hunk 1]
```diff
@@ -1,32 +1,48 @@
...
-# Build the google test library.
-$(LIBGTEST): $(LIBGTEST)(test/gtest.o)
-
-test/gtest.o: $(GTEST)/src/gtest-all.cc
-	@mkdir -p test
-	$(COMPILE.cc) -O$O $(GTEST_CXXFLAGS) $< $(OUTPUT_OPTION)
+test/%$(EXE) : CXXFLAGS += $(CXXFLAGS_GTEST)
+test/%$(EXE) : CPPFLAGS += $(CPPFLAGS_GTEST)
+test/%$(EXE) : INC_FIRST = -I $(if $(STAN),$(STAN)/src,src) -I $(if $(STAN),$(STAN),.)
+test/%$(EXE) : INC += $(INC_GTEST)
```
Comment (seantalts) [inline make/tests]: I'm trying to use these for the first time and getting pretty turned around... Where is `INC` used, if anywhere?

I'm trying to figure out how to compile something that includes the correct Stan includes.

**Pull request merged**